server/main_server/src/agent_main.cpp

```diff
@@ -0,0 +1,1326 @@
+                status = sslAccept(&_comm);
+                if (status < 0) {
+                    log_agent::error("sslAccept failed in agentMain with status {}", status);
+                }
+                _comm.ssl_do_accept = 0;
+            }
+            if (_comm.ssl_do_shutdown) {
+                status = sslShutdown(&_comm);
+                if (status < 0) {
+                    log_agent::error("sslShutdown failed in agentMain with status {}", status);
+                }
+                _comm.ssl_do_shutdown = 0;
+            }
+
+            status = readAndProcClientMsg(&_comm, READ_HEADER_TIMEOUT);
+            if (status < 0) {
+                if (status == DISCONN_STATUS) {
+                    status = 0;
+                    break;
+                }
+            }
+        }
+
+        irods::experimental::api::plugin_lifetime_manager::destroy();
+
+        // determine if we even need to connect, break the
+        // infinite reconnect loop.
+        if (!resc_mgr.need_maintenance_operations()) {
+            return status;
+        }
+
+        // find the icat host
+        rodsServerHost_t* rodsServerHost = 0;
+        status = getRcatHost(PRIMARY_RCAT, 0, &rodsServerHost);
+        if (status < 0) {
+            log_agent::error(ERROR(status, "getRcatHost failed.").result());
+            return status;
+        }
+
+        // connect to the icat host
+        status = svrToSvrConnect(&_comm, rodsServerHost);
+        if ( status < 0 ) {
+            log_agent::error(ERROR(status, "svrToSvrConnect failed.").result());
+            return status;
+        }
+
+        // call post disconnect maintenance operations before exit
+        status = resc_mgr.call_maintenance_operations(rodsServerHost->conn);
+
+        return status;
+    } // agentMain
+
+    auto create_agent_pid_file_for_ips(const RsComm& _comm, std::time_t _created_at) -> void
+    {
+        std::string_view local_zone = "UNKNOWN";
+        if (const auto* lz = getLocalZoneName(); lz) {
+            local_zone = lz;
+        }
+
+        std::string_view client_zone = _comm.clientUser.rodsZone;
+        if (client_zone.empty()) {
+            client_zone = local_zone;
+        }
+
+        std::string_view proxy_zone = _comm.proxyUser.rodsZone;
+        if (proxy_zone.empty()) {
+            proxy_zone = local_zone;
+        }
+
+        std::string_view client_program_name = _comm.option;
+        if (client_program_name.empty()) {
+            client_program_name = "UNKNOWN";
+        }
+
+        const auto proc_dir = irods::get_irods_proc_directory();
+        const auto pid_file = fmt::format("{}/{}", proc_dir.c_str(), getpid());
+
+        if (std::ofstream out{pid_file}; out) {
+            out << fmt::format("{} {} {} {} {} {} {}\n",
+                _comm.clientUser.userName,
+                client_zone,
+                _comm.proxyUser.userName,
+                proxy_zone,
+                client_program_name,
```

Comment:
In iRODS 4, the proxy user and client user are swapped. Based on what I've seen in rsProcStat, that doesn't seem correct.

Need someone else to verify this in iRODS 4.

Reply:
Created #8001

Reply:
After looking at the implementation of `ips` and the functions in the server which support it, I'm sure the highlighted code is correct and the iRODS 4 implementation has a bug.

Resolving this review comment since there's nothing to do.